Feat/jekyll4

gweiying opened the pull request:
This PR merges the bash scripts for Jekyll 4 migration into `isomer-conversion-scripts` staging. 

## Details

This PR introduces 4 key changes:

1) Replaces GitHub Pages gem with `isomer-jekyll` gem. Adds Isomer-specific gem dependencies, modifies `plugins:` in `_config.yml`
2) Centralizes `build.sh` and `netlify.toml` files, retrieved at build time from [isomer-build](https://github.com/opengovsg/isomer-build)
3) Modifies collection structure, introduces `.keep` files, removes `collections` specifications from `_config.yml`
4) Adds staging and prod Netlify urls to `_config.yml`

## Background

We have decided to migrate Isomer sites to Jekyll 4.0 to use its native reordering capabilities. The above changes have been implemented in the migration scripts [here](https://github.com/isomerpages/isomer-conversion-scripts/tree/feat/jekyll4). This PR updates site-creation-backend to generate sites with the Jekyll 4, CMS-compatible structure.

#### Replacing Github Pages gem

We have decided to replace the GitHub Pages gem with our own `isomer-jekyll` gem as we would like to upgrade Jekyll to 4.0 (which GitHub Pages' gem does not support), and also because maintaining our own `isomer-jekyll` gem would allow us more control over gem dependencies and pushing out updates to Isomer repos. More information on our rationale for maintaining our own `isomer-jekyll` gem can be found [here](https://docs.google.com/document/d/1ZVSgFhNClGL0K9Q8udePnwftPJ1Qex61Y0QG8QOXM9A/edit#)

#### Centralizing `build.sh` and `netlify.toml` files

We have decided to centralise our `build.sh` and `netlify.toml` files, and host it here ([isomer-build](https://github.com/opengovsg/isomer-build)) to be retrieved at build time. This will prevent any unintentional or malicious script and CSP injections from our users. More information on our rationale for maintaining centralised `build.sh` and `netlify.toml` files can be found [here](https://docs.google.com/document/d/1ZVSgFhNClGL0K9Q8udePnwftPJ1Qex61Y0QG8QOXM9A/edit#)

#### Modifying collection structure and introduces `.keep` files to subfolders

We introduce a new collection structure, with three key changes:

Each collection will contain a directory file: collection.yml, which will store the current order of files, and a boolean value for whether the collection should be included as part of the deployed site, output: true. This collection.yml follows the format of a Jekyll _config.yml file, and acts as an extension to the configurations provided in <repo>/_config.yml. A sample collection.yml is as follows:
```
collections:
  about-us:
    output: true
    order:
      - mission/.keep
      - mission/goals.md
      - mission/statement.md
      - contacts.md
```
Collections with third-nav files will now have explicit nesting folders for these third-nav files. While this change does not change any current functionality of Isomer sites, we introduce this in preparation for refactoring Isomer templates to remove the third_nav_title metatag. An example of the new collection structure is as follows:
```
_about-us/
   - collection.yml
   - mission
     - goals.md
     - statement.md
     - .keep
   - contacts.md
```
The collection.yml files extend the configurations provided in <repo>/_config.yml and have to be provided to the Jekyll build command under the --config option when the site is built, for example JEKYLL_ENV=staging jekyll build --config _config.yml,_about-us/collection.yml. This is handled at build time by our `build.sh` file, hosted at [isomer-build](https://github.com/opengovsg/isomer-build).

Finally, we also introduce a .keep file within subfolders, to support the creation of empty subfolders on the CMS. These .keep files are also added to subfolders in `/images` and `/files`.

#### Adds staging and prod Netlify urls to `_config.yml`
This change is made so that the CMS will be able to retrieve the staging and prod urls for a url.

gweiying committed:
feat: add jekyll 4 migration files

gweiying committed:
fix: improve error handling, handles empty collections

gweiying committed:
fix: remove migration branch if error

gweiying committed:
fix: removes old Gemfile and creates new Gemfile from scratch

gweiying committed:
fix: removes entire Gemfile instead of deleting specific lines from Gemfile

gweiying committed:
fix: adds code to parse config file for collection output values, and uses that to populate new collections.yml files

gweiying committed:
Update README.md

gweiying committed:
Update README.md

gweiying committed:
Update README.md

gweiying committed:
fix: add comment for substring function

LoneRifle commented:
are you working on something similar for sites created via site-creation-backend?

gweiying committed:
fix: deploy.sh bug on Netlify

Previously, the deployment script compiled a command to find all the collection.yml files in a site repository, and concatenate them into a command via a loop. However, since the default internal field separator IFS was spaces, collection.yml file paths that contained spaces, e.g. /_hello new folder/collection.yml could not be parsed as one filepath, but was parsed separately. A hacky fix was introduced to temporarily change the IFS to new lines, complete the loop, then set the IFS back to the default. This fix worked locally, but introduced a bug into the Netlify deployment, as the GNU shell does not seem to be compatible with this IFS change trick.

This PR replaces the loop with a replacement command to change from space-separated collection.yml files to comma-separated files, which Jekyll accepts as the argument to --config.

An example of Netlify deployment bug can be found here https://app.netlify.com/sites/friendly-pare-41cc65/deploys/6036fcd1c0fbf600070b9021, line 63.

gweiying committed:
Update README.md

gweiying committed:
Update README.md

gweiying committed:
Update README.md

alexanderleegs committed:
Feat: add placeholder file during migration (#18)

gweiying committed:
fix: remove deploy.sh from migration steps

gweiying committed:
fix: modify repo with isomer-jekyll gem file-related changes

gweiying committed:
fix: rename function for removing _config.yml collection blurb

gweiying committed:
fix: add function for updating _config.yml plugins blurb

gweiying committed:
fix: rename migration scripts for clarity

gweiying committed:
fix: update README.md

gweiying committed:
fix: error in git file tracking after gem update

alexanderleegs committed:
feat: add netlify urls to _config.yml

alexanderleegs committed:
feat: add placeholder file to nested image/file directories

alexanderleegs committed:
Update README.md

gweiying committed:
fix: add check for empty collection folders and pages with no frontmatter

gweiying committed:
fix: add check for md pages outside of /pages

alexanderleegs committed:
Feat/netlify url in config and placeholder files for images (#20)

* feat: add netlify urls to _config.yml

* feat: add placeholder file to nested image/file directories

* Update README.md

gweiying committed:
fix: typo in migration.sh

gweiying committed:
Merge branch 'feat/isomer-gem-update' into fix/additional-repo-checks

gweiying committed:
Merge pull request #21 from isomerpages/fix/additional-repo-checks

Fix/additional repo checks

gweiying committed:
Merge pull request #19 from isomerpages/feat/isomer-gem-update

Feat/isomer-jekyll-update

gweiying committed:
fix: update README.md

alexanderleegs reviewed (approved):
lgtm

alexanderleegs committed:
Fix: remove quotes from retrieved repo description

gweiying committed:
Merge pull request #23 from isomerpages/fix-remove-quotes-from-retrieved-description

Fix: remove quotes from retrieved description

alexanderleegs committed:
Fix/handle empty collections and commit placeholder image files (#24)

* Fix: handle case where no collections exist

* Fix: handle case where no nested images exist

alexanderleegs committed:
Fix: handle case where no collection in config exist

alexanderleegs committed:
Fix: update env example for resource rename branch ref env var

gweiying committed:
Merge pull request #25 from isomerpages/fix-handle-empty-collections

Fix: handle case where no collection in config exist

gweiying committed:
fix: conditionally replacing 4 space with 2 space in _config.yml (#27)

alexanderleegs committed:
Fix/retain resource layout and contact us fix (#31)

* Fix: Retain post layout

* Fix: handle empty line on contact us

* Fix: remove console.log statement

* Fix: filter out null values in script instead of in checker

* Fix: swap order of filter and map
